pkg/cmd/repo/sync/sync.go

```diff
@@ -0,0 +1,286 @@
+
+	if opts.DestArg == "" {
+		local = true
+		destRepo, err = opts.BaseRepo()
+		if err != nil {
+			return err
+		}
+	} else {
+		destRepo, err = ghrepo.FromFullName(opts.DestArg)
+		if err != nil {
+			return err
+		}
+	}
+
+	if opts.SrcArg == "" {
+		if local {
+			srcRepo = destRepo
+		} else {
+			opts.IO.StartProgressIndicator()
+			srcRepo, err = api.RepoParent(apiClient, destRepo)
+			opts.IO.StopProgressIndicator()
+			if err != nil {
+				return err
+			}
+			if srcRepo == nil {
+				return fmt.Errorf("can't determine source repo for %s because repo is not fork", ghrepo.FullName(destRepo))
+			}
+		}
+	} else {
+		srcRepo, err = ghrepo.FromFullName(opts.SrcArg)
+		if err != nil {
+			return err
+		}
+	}
+
+	if !local && destRepo.RepoHost() != srcRepo.RepoHost() {
+		return fmt.Errorf("can't sync repos from different hosts")
+	}
+
+	if opts.Branch == "" {
+		opts.IO.StartProgressIndicator()
+		opts.Branch, err = api.RepoDefaultBranch(apiClient, srcRepo)
+		opts.IO.StopProgressIndicator()
+		if err != nil {
+			return err
+		}
+	}
+
+	srcStr := fmt.Sprintf("%s:%s", ghrepo.FullName(srcRepo), opts.Branch)
+	destStr := fmt.Sprintf("%s:%s", ghrepo.FullName(destRepo), opts.Branch)
+	if local {
+		destStr = fmt.Sprintf(".:%s", opts.Branch)
+	}
+	cs := opts.IO.ColorScheme()
+	if !opts.SkipConfirm && opts.IO.CanPrompt() {
+		if opts.Force {
+			fmt.Fprintf(opts.IO.ErrOut, "%s Using --force will cause diverging commits on %s to be discarded\n", cs.WarningIcon(), destStr)
+		}
+		var confirmed bool
+		confirmQuestion := &survey.Confirm{
+			Message: fmt.Sprintf("Sync %s from %s?", destStr, srcStr),
+			Default: false,
+		}
+		err := prompt.SurveyAskOne(confirmQuestion, &confirmed)
+		if err != nil {
+			return err
+		}
+
+		if !confirmed {
+			return cmdutil.CancelError
+		}
+	}
+
+	opts.IO.StartProgressIndicator()
+	if local {
+		err = syncLocalRepo(srcRepo, opts)
+	} else {
+		err = syncRemoteRepo(apiClient, destRepo, srcRepo, opts)
+	}
+	opts.IO.StopProgressIndicator()
+
+	if err != nil {
+		return err
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		success := cs.Bold(fmt.Sprintf("Synced %s from %s\n", destStr, srcStr))
+		fmt.Fprintf(opts.IO.Out, "%s %s", cs.SuccessIconWithColor(cs.GreenBold), success)
+	}
+
+	return nil
+}
+
+func syncLocalRepo(srcRepo ghrepo.Interface, opts *SyncOptions) error {
+	// Remotes precedence by name
+	// 1. upstream
+	// 2. github
+	// 3. origin
+	// 4. other
+	remotes, err := opts.Remotes()
+	if err != nil {
+		return err
+	}
+	remote := remotes[0]
+	branch := opts.Branch
+
+	_ = executeCmds([][]string{{"git", "fetch", remote.Name, fmt.Sprintf("+refs/heads/%s", branch)}})
+
+	hasLocalBranch := git.HasLocalBranch(branch)
+	if hasLocalBranch {
+		fastForward, err := git.IsAncestor(branch, fmt.Sprintf("%s/%s", remote.Name, branch))
+		if err != nil {
+			return err
+		}
+
+		if !fastForward && !opts.Force {
+			return fmt.Errorf("can't sync .:%s because there are diverging commits, try using `--force`", branch)
```

Comment:
I'm not sure if this is exactly right and might be too verbose, but I'm wary of just suggesting --force without any info about its destructive behavior.
```suggestion
			return fmt.Errorf("can't sync .:%s because there are diverging commits, you can use `--force` to overwrite the commits on .:%s", branch)
```